### PR TITLE
R11s: log response finish-to-close gap

### PR DESCRIPTION
**Pull request by znewton**
## Motivation

In AFR, we get a lot of reports about slow response times even when our service telemetry reports more reasonable response times. This can range from a couple hundred milliseconds to a minute or more. From a service reliability perspective, we _need_ to get visibility there. This is an **experimental** change to attempt to capture part of that story in telemetry.

## Description

Node.js Responses have 2 events: `close` and `finish`. According to Node.js documentation:

> Event: 'close' [#](https://nodejs.org/docs/latest-v16.x/api/http.html#event-close_2)
> Added in: v0.6.7
> 
> Indicates that the response is completed, or its underlying connection was terminated prematurely (before the response completion).

> Event: 'finish' [#](https://nodejs.org/docs/latest-v16.x/api/http.html#event-finish_1)
> Added in: v0.3.6
> 
> Emitted when the response has been sent. More specifically, this event is emitted when the last segment of the response headers and body have been handed off to the operating system for transmission over the network. It does not imply that the client has received anything yet.

[morgan](https://github.com/expressjs/morgan), the Express middleware we use for logging HTTP Request duration, logs on the 'finish' event (or 'close' if premature termination) using the [on-finished](https://www.npmjs.com/package/on-finished) package.

The goal here is to log the gap between request "finish" (hand-off to OS) and "close" (response complete).

## Caveats

I don't actually know if this will work as intended. There are some gaps in my knowledge about what is actually happening here. However, better to have a little more telemetry than not, as we suspect Network I/O could be causing bottlenecks in AFR.

### CODE REVIEW

**Comment by znewton** [comment]
This did not work

**Comment by znewton** [comment]
Re-opening with a new intention: finish->close is _**not**_ representational of time to respond to the client. However, it _could_ be representational of time from sending response to OS and OS closing the connection. Still not sure, but I think this could be helpful regardless.

File: server/routerlicious/packages/services-utils/src/morganLoggerMiddleware.ts [hunk 1]
```diff
@@ -34,6 +34,24 @@ export function alternativeMorganLoggerMiddleware(loggerFormat: string) {
...
+	 * More specifically, this event is emitted when the last segment of the response headers and body have been handed off to the operating system for transmission over the network.
+	 * It does not imply that the client has received anything yet.
+	 * Docs: https://nodejs.org/docs/latest-v16.x/api/http.html#event-finish_1
+	 */
+	finishTime: number | undefined;
+	/**
+	 * Indicates that the response is completed, or its underlying connection was terminated prematurely (before the response completion).
+	 * Docs: https://nodejs.org/docs/latest-v16.x/api/http.html#event-close_2
+	 */
+	closeTime: number;
```
Comment (pradeepvairamani) [inline server/routerlicious/packages/services-utils/src/morganLoggerMiddleware.ts:52]: Can you please help me validate my understanding:

When the request is handled successfully and there is no timeout: both close and finish will be called with finish having a very small time delay compared to when close is emitted or does finish come before close?

If it is prematurely terminated, only close will be called.

If the request errored out but it is not the underlying connection being terminated, both finish and close will be called?

**Review by pradeepvairamani — APPROVED** [review]

File: server/routerlicious/packages/services-utils/src/morganLoggerMiddleware.ts (hunk 1 above)
Comment (znewton) [reply server/routerlicious/packages/services-utils/src/morganLoggerMiddleware.ts:52]: `finish` comes before `close`, normally with a very small time difference.

Correct that if the request is prematurely ended (e.g. by client due to timeout) then only `close` will be called.

I believe, on error, `error` would happen, then `close`, with no `finish`. It's also possible that there would be `error` but nothing else, which is why I added the `reject on error` code to be safe.

Comment (znewton) [reply server/routerlicious/packages/services-utils/src/morganLoggerMiddleware.ts:52]: There is also an event in the parent class of ServerResponse, OutgoingMessage, called [prefinish](https://nodejs.org/docs/latest-v20.x/api/http.html#event-prefinish), which might be worth logging here.

**Pull request merged**